File: compiler/bench/isprime.fs
\ expected: 78498
\ Primality test, 1M numbers
\ Checksum: count of primes found

: is-prime ( n -- 0|1 )
  dup 2 < if drop 0 exit then
  dup 2 = if drop 1 exit then
  dup 2 mod 0= if drop 0 exit then
  dup 3 = if drop 1 exit then
  dup 3 mod 0= if drop 0 exit then
  5
  begin
    2dup dup * >= while
    2dup mod 0= if 2drop 0 exit then
    2 +
    2dup mod 0= if 2drop 0 exit then
    4 +
  repeat 2drop 1 ;

: bench-isprime ( -- count )
  0
  1000000 0 do
    i is-prime +
  loop ;

: main bench-isprime . cr ;

File: compiler/bench/isprime.c
// expected: 78498
// Primality test, 1M numbers
// Checksum: count of primes found

#include <stdio.h>

int is_prime(long n) {
    if (n < 2) return 0;
    if (n == 2) return 1;
    if (n % 2 == 0) return 0;
    if (n == 3) return 1;
    if (n % 3 == 0) return 0;
    long i = 5;
    while (i * i <= n) {
        if (n % i == 0) return 0;
        i += 2;
        if (n % i == 0) return 0;
        i += 4;
    }
    return 1;
}

int main(void) {
    long count = 0;
    for (long i = 0; i < 1000000; i++) {
        count += is_prime(i);
    }
    printf("%ld\n", count);
    return 0;
}
